common/src/main/java/com/microsoft/identity/common/internal/fido/AuthFidoChallengeHandler.kt

```diff
@@ -0,0 +1,272 @@
+                FidoRequestField.RELYING_PARTY_IDENTIFIER.fieldName,
+                fidoChallenge.relyingPartyIdentifier
+            )
+            userVerificationPolicy = validateRequiredParameter(
+                FidoRequestField.USER_VERIFICATION_POLICY.fieldName,
+                fidoChallenge.userVerificationPolicy
+            )
+            allowedCredentials = validateOptionalListParameter(
+                AuthFidoRequestField.ALLOWED_CREDENTIALS.fieldName,
+                fidoChallenge.allowedCredentials
+            )
+            validateProtocolVersion(fidoChallenge.version)
+            //Not currently using keyTypes, but should validate in case we use it in the future.
+            validateOptionalListParameter(
+                AuthFidoRequestField.KEY_TYPES.fieldName,
+                fidoChallenge.keyTypes
+            )
+        } catch (e : Exception) {
+            respondToChallengeWithError(
+                submitUrl = submitUrl,
+                context = context,
+                errorMessage = e.message.toString(),
+                exception = e,
+                methodTag = methodTag
+            )
+            return null
+        }
+        if (lifecycleOwner == null) {
+            respondToChallengeWithError(
+                submitUrl = submitUrl,
+                context = context,
+                errorMessage = "Cannot get lifecycle owner needed for FIDO API calls.",
+                methodTag = methodTag
+            )
+            return null
+        }
+        lifecycleOwner.lifecycleScope.launch {
+            try {
+                val assertion = fidoManager.authenticate(
+                    challenge = authChallenge,
+                    relyingPartyIdentifier = relyingPartyIdentifier,
+                    allowedCredentials = allowedCredentials,
+                    userVerificationPolicy = userVerificationPolicy
+                )
+                span.setStatus(StatusCode.OK)
+                respondToChallenge(
+                    submitUrl = submitUrl,
+                    assertion = assertion,
+                    context = context
+                )
+            } catch (e : Exception) {
+                respondToChallengeWithError(
+                    submitUrl = submitUrl,
+                    context = context,
+                    errorMessage = e.message.toString(),
+                    exception = e,
+                    methodTag = methodTag
+                )
+            }
+        }
+        return null
+    }
+
+    /**
+     * Validates that the given required parameter is not null or empty.
+     * @param field passkey protocol field
+     * @param value value for a passkey protocol parameter
+     * @return validated parameter value
+     * @throws ClientException if the parameter is null or empty.
+     */
+    @Throws(ClientException::class)
+    fun validateRequiredParameter(field: String, value: String?): String {
```

Comment:
you can do something like this

```Kotlin
        data class FidoChallengeField<K>(private val field: FidoRequestField,
                                         private val value: K?,
                                         private val throwIfValueIsInvalid: (FidoRequestField, K) -> Void) {
            
            @Throws(ClientException::class)
            fun getOrThrow(): K {
                if (value == null){
                    throw ClientException("$field is null")
                }

                throwIfValueIsInvalid(value)
                return value
            }
            
            companion object {
                @Throws(ClientException::class)
                fun throwIfInvalidUrl(field: FidoRequestField,
                                      url: String) {
                    try {
                        URL(url)
                    } catch (e : MalformedURLException) {
                        throw ClientException(
                            ClientException.PASSKEY_PROTOCOL_REQUEST_PARSING_ERROR,
                            "$field value is not a valid URL.")
                    }
                }
            }
        }

        data class FidoChallenge(
            val submitUrl: FidoChallengeField<String>,
            ...) {

            constructor(redirectUri: String) : this(
                 submitUrl = FidoChallengeField<String>(
                    FidoRequestField.SUBMIT_URL,
                    parameters[FidoRequestField.SUBMIT_URL.fieldName],
                    ::throwIfInvalidUrl
                ),
                ...
            )
        }
```
then
``` Kotlin
    fidoChallenge.submitUrl.getOrThrow()
```

I'll let you decide if you want to go that route.

Reply:
Took your suggestion: 40f33a68e5d35e99684e0f93679acb9135abe05b